libgnucash/engine/ScrubBudget.c

```diff
@@ -0,0 +1,238 @@
+/********************************************************************\
+ * ScrubBudget.c -- fix budget amount signs                         *
+ *                                                                  *
+ * This program is free software; you can redistribute it and/or    *
+ * modify it under the terms of the GNU General Public License as   *
+ * published by the Free Software Foundation; either version 2 of   *
+ * the License, or (at your option) any later version.              *
+ *                                                                  *
+ * This program is distributed in the hope that it will be useful,  *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
+ * GNU General Public License for more details.                     *
+ *                                                                  *
+ * You should have received a copy of the GNU General Public License*
+ * along with this program; if not, contact:                        *
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+ *                                                                  *
+\********************************************************************/
+
+#include <config.h>
+
+#include <glib.h>
+#include <glib/gi18n.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+
+#include "gnc-budget.h"
+#include "gnc-features.h"
+#include "ScrubBudget.h"
+
+static QofLogModule log_module = "gnc.engine.scrub";
+
+typedef enum
+{
+    HEURISTICS_INC_EXP,
+    HEURISTICS_CREDIT_ACC,
+    HEURISTICS_NONE
+} HeuristicsOutput;
+
+typedef struct
+{
+    GncBudget* budget;
+    HeuristicsOutput policy;
+} Reversal_t;
+
+typedef struct
+{
+    gint asset,liability,equity,income,expense;
+    gint num_periods;
+    GncBudget* budget;
+} Heuristics;
+
+static gboolean this_book_needs_fixing;
+
+static void process_heuristics_acct (Account * account, gpointer user_data)
+{
+    /* each account- check type. sum budget period amounts. if sum<0,
+       decrease type tally by 1. if sum>0, increase type tally by 1. */
+    Heuristics *heuristics = (Heuristics*) user_data;
+    gnc_numeric total = gnc_numeric_zero(), val;
+    gint i;
+
+    for (i = 0; i < heuristics->num_periods; ++i)
+    {
+        if (!gnc_budget_is_account_period_value_set (heuristics->budget, account, i))
+            continue;
+        val = gnc_budget_get_account_period_value (heuristics->budget, account, i);
+        total = gnc_numeric_add_fixed (total, val);
+    }
+
+    i = gnc_numeric_compare (total, gnc_numeric_zero ());
+    PWARN ("acct=%s, total=%ld/%ld, i=%d", xaccAccountGetName (account),
```

Comment:
You can't count on `int64_t` mapping to `long`, so instead use either the standard C `PRId64` or the GLib equivalent `G_GINT64_FORMAT`  :
```
"acct=%s, total=%" PRId64 "/%" PRId64 ", i=%d",
```

Reply:
`PRId64` errors out. I'll change to `gnc_numeric_to_string`

Reply:
You might need to `#include inttypes.h>` to get `PRId64`. `gnc_numeric_to_string` uses `G_GINT64_FORMAT`, but it also allocates so you'll need to assign the string to a variable that you can free after the message.